assets/js/base/components/checkbox-list/index.js

```diff
@@ -0,0 +1,190 @@
+/**
+ * External dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import PropTypes from 'prop-types';
+import {
+	Fragment,
+	useCallback,
+	useMemo,
+	useState,
+	useEffect,
+} from '@wordpress/element';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+
+/**
+ * Component used to show a list of checkboxes in a group.
+ */
+const CheckboxList = ( {
+	className,
+	onChange = () => {},
+	options = [],
+	isLoading = false,
+	limit = 10,
+} ) => {
+	// Holds all checked options.
+	const [ checked, setChecked ] = useState( [] );
+	const [ showExpanded, setShowExpanded ] = useState( false );
+
+	useEffect( () => {
+		onChange( checked );
+	}, [ checked ] );
+
+	const placeholder = useMemo( () => {
```

Comment:
As I mentioned in my previous comment:

> Advantage of using useMemo here is that it only calculates the value once when this component is mounted. Thus any additional calls to the function on lifecycle changes while the component is mounted will return the memoized value instead of regenerating the list items.

Which would you prefer?

No `useMemo`:

<img width="1431" alt="Image 2019-11-07 at 11 43 03 AM" src="https://user-images.githubusercontent.com/1429108/68408859-11754480-0154-11ea-9c09-e25667c88e0c.png">

With `useMemo`:

<img width="1422" alt="Image 2019-11-07 at 11 43 51 AM" src="https://user-images.githubusercontent.com/1429108/68408880-1afeac80-0154-11ea-9ec7-6d6a0917a51e.png">

I added `console.log( 'calculating...' )` whenever placeholder is invoked for the above.

Reply:
Yeah I was coming back to say this; it re-renders the placeholder many times. `useMemo` does not so stick with that.

Reply:
Incidentally, understanding what is happening in react with functional components (and now with Hooks) was one of the hardest things for me to wrap my head around (and even now I still sometimes forget and get it wrong). When in doubt, I'll drop in some console.logs and watch what happens :)

Reply:
You're right, I'm new to hooks so it's still not clear to me when it's worth using `useMemo()` and when it's not. Technically, we could also wrap the [classes assignment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1127/files#diff-53d339601cf4e1728e109362c7ca6207R159) into a `useMemo()` function, right? But that would probably be over-kill since it's not an expensive function at all. Creating the `placeholder` seems a bit more expensive, but when is a function "expensive enough" to require `useMemo()`? I wonder if you have a rule of thumb to know when it's worth using `useMemo()` and when it's not.

Regarding this specific case, why not moving `placeholder`'s definition outside of the `CheckboxList` component so it's only computed once without the need of `useMemo()`? There might still be the edge case in which we import `chebox-list/index.js` but don't create `CheckboxList`, in that case we would be creating the `placeholder` but not using its value.  But I think we can neglect this case.

Reply:
> Technically, we could also wrap the classes assignment into a useMemo() function, right?

Right.  But as you pointed out, it's overkill because it's not expensive.

> but when is a function "expensive enough" to require useMemo()? I wonder if you have a rule of thumb to know when it's worth using useMemo() and when it's not.

No rule of thumb, and likely the performance gain with adding `useMemo` here is not huge.  But for things like a placeholder list I think it doesn't hurt.

> Regarding this specific case, why not moving placeholder's definition outside of the CheckboxList component so it's only computed once without the need of useMemo()?

Yup, anytime there are no dependencies, it's a good candidate to simply define it as a variable outside of the function. But there's also advantages to keeping things defined in the component using them when it comes to code review, and context.